docs/rules/no-html-safe.md

```diff
@@ -0,0 +1,112 @@
+# no-html-safe
+
+`htmlSafe` marks a string as safe for unescaped output with Ember templates so you can render it as HTML. While useful this can inadvertently open you up to Cross-site Scripting (XSS) vulnerabilities, especially if the string was generated from user input or some other untrusted source. **You should only ever use `htmlSafe` with trusted or sanitized input**.
+
+## Rule Details
+
+This rule prevents importing the `htmlSafe` utillity from `@ember/template` (or `@ember/string` for older Ember versions);
+
+## Examples
+
+Examples of **incorrect** code for this rule:
+
+```js
+import { htmlSafe } from '@ember/template';
+```
+
+```js
+import { htmlSafe } from '@ember/string';
+```
+
+## Alternatives
+
+There are a few alternative strategies to using `htmlSafe`.
+
+### Scenario
+
+In this example we have a property `userContent` which contains a user-generated string of HTML. For example, something like `<h1>Joe Bloggs</h1>`. This is then rendered directly into the template.
+
+``` js
+// components/my-component.js
+import { htmlSafe } from '@ember/template';
+
+class MyComponent extends Component {
+  get myContent() {
+    return htmlSafe(this.args.userContent);
+  }
+}
+```
+
+```hbs
+{{!-- components/my-component.hbs --}}
+{{this.myContent}}
+```
+
+### Alternative 1: Render the HTML yourself
+
+While not as flexible, if you can control the content generated by the user you should only let the user enter plaintext and render the HTML yourself.
+
+```hbs
+{{!-- components/my-component.hbs --}}
+<h1>{{@userContent}}</h1>
+```
+
+### Alternative 2: Define your own trusted sanitizing helper
```

Comment:
Regarding alternatives 2 and 3, I'm unclear why it would be better to define a custom sanitizing helper. In fact, encouraging custom sanitizing helpers sound like a dangerous anti-pattern, as they would be much more likely to have bugs / security vulnerabilities versus using a published/trusted helper like `htmlSafe` from Ember.

Reply:
So htmlSafe doesn't do _any_ sanitization which is why it can easily become a XSS vector. With a name like htmlSafe I've been caught out and reviewed many PRs where people assume it is doing sanitization. All it does is takes the string with the HTML in it and says "yep you can render this string in a template as html", even if it has XSSs in it.

React has a similar helper to htmlSafe but it's called dangerouslySetInnerHtml which is a lot more obvious you shouldn't be using it. They even have a [similar lint rule](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-danger.md).

For 2 and 3, I'm advocating if you have to use htmlSafe always sanitize the content first that you are going to mark safe. I'd recommend a sanitizer like DOMPurify but wasn't sure whether I'd call it out.

Happy to get further feedback though. Even happy to cut back the section altogether if you think it will just add to confusion.

Reply:
I've tweaked the opening paragraph mentioning that `htmlSafe` doesn't perform sanitization, and also added updated alternative to mention DOMPurify.

Reply:
Ah my bad. I knew `htmlSafe` doesn't do any sanitizing but it's easy get confused about it due to its poor name as you mention.

This plan sounds good:
* Updating the doc to help people avoid that confusion
* Suggesting something like DOMPurify so people don't think they need to write a custom sanitizer with their own string parsing

Reply:
Ya, the name really is unfortunate. It is _trying_ to be the developer saying "hey this is html safe", but most people read it as "hey please make this html safe for me".